check if User exists before saving UserSessionMembership

fosterseth opened the pull request:
##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->
related to issue #4334 

Change: check that User exists before saving the UserSessionMembership

Significant line addition: `if User.objects.filter(id=int(user_id)).exists():`

An internal database error occurs if
1. User is logged in with active session
2. User is deleted
3. User reloads browser, or makes an HTTP API request

Error stems from signals.py function save_user_session_membership()

UserSessionMembership model will grab the user_id from the session (which now contains stale info) and attempt to save to the database using User as a foreignkey. However, because User was already deleted, it will throw the error.

The solution is to check that a User exists with the same user_id as the session user_id.

<!---
If you are fixing an existing issue, please include "related #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Bugfix Pull Request

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
 - API

##### AWX VERSION
<!--- Paste verbatim output from `make VERSION` between quotes below -->
```
awx: 7.0.0
```


##### ADDITIONAL INFORMATION
<!---
Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful.
  -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```


```

fosterseth committed:
Saving user session checks if User exists

- Check that model User object exists with id=user_id
  before attempting to save to database
- UserSessionMembership saves to the database using
  foreign key, User
- However, User with matching id might not exist if
  browser sends request with stale cookies
- Change made in regards to issue #4334

fosterseth committed:
removed duplicate import of User

ryanpetrello reviewed (approved):
Seems pretty reasonable to me 👍

AlanCoding commented on awx/main/signals.py:
```diff
@@ -684,7 +684,9 @@ def save_user_session_membership(sender, **kwargs):
-    UserSessionMembership(user_id=user_id, session=session, created=timezone.now()).save()
+    # check if user_id from session has an id match in User before saving
+    if User.objects.filter(id=int(user_id)).exists():
+        UserSessionMembership(user_id=user_id, session=session, created=timezone.now()).save()
```
seems like you should go ahead and indent the rest of this method. If the user doesn't exist, I don't think we want to go expiring session memberships for that deleted user.

fosterseth committed:
check expired sessions only if User exists

- Indent rest of code into the conditional
  that checks for expired sessions of that User
- If user doesn't exist, no need to check

AlanCoding reviewed (approved):
I think this looks fine to me, but maybe @rooftopcellist should take a look too

rooftopcellist commented:
@fosterseth This looks good.  I tested it out locally and it now gives back a 401 instead of a 500.  Interestingly, after it gets back the first 401, the UI continues to hit ~4 other endpoints, consequently getting denied 4 more times.  If there is a quick and dirty way to avoid this, we should do it; otherwise I am OK with merging this as is.  

For ref:
![image](https://user-images.githubusercontent.com/11698892/65990755-2aeee680-e45a-11e9-9b9c-b9ce0e96b7d0.png)

ryanpetrello commented:
@rooftopcellist I'm fine w/ this new behavior.
